code/headers/mq_sensors_gas_curve_c.hpp

```diff
@@ -0,0 +1,48 @@
+#pragma once
+#include <gas_enums.hpp>
+#include <array>
+
+namespace r2d2::gas_detection {
+
+    /**
+     * Class mq_sensors_gas_curve_c provides the ability to create gas_curves
+     * for the mq_sensor. Each gas has its own logarithmic curve. The first 2
+     * values are points on the curve, the 3rd value is the slope increment.
+     */
+    class mq_sensors_gas_curve_c {
+    private:
+        int gas_id;
+        std::array<float, 3> gas_curve = {};
```

Comment:
Nitpick maybe, but why is this an array if you're storing 2 semantically 'different' tings? 
I think an array of two points on the curve and a variable containing the slope would be better.

Reply:
I changed every thing you requested except this request. 
The reason why you would keep them together is because it's just one thing. 
Don't see it as an x, y, slope but as **one curve**. For that reason, the curves are displayed as 3 floats. but good notice!

Reply:
I don't see the changes implemented yet, so I can't yet resolve the comments I've placed. Make sure to merge them into release.

I've consulted with the code-reviewers about this specific issue, the thing is, this class should be a POD struct, since it doesn't offer any functionality other than storage of a few values and an ID. 

These are the two options we'd be happy with, keeping in mind the fact that this project will have to be maintained by a new R2D2-team next year.

1. Document the array thoroughly using doxygen, that way the new teams will at least be able to continue using this class.
2. Change the array to only hold the `x` and `y` value and add a `float curve;` member. This would be the most elegant solution.

The reason for the comment is that the class itself describes and "contains" *one curve*. The array is just a member of the class. 
The fact that the curves are displayed as 3 floats is fine, but the way they are stored is not logical. 

We'll leave it up to you and your team to see which option you opt to go for.

Reply:
Hi,
After receiving a message about this and since I created this class, I think I can give an explanation about the choices I made here.

- The main reason why I chose this to be a class and not a struct is because it has a constructor. It also has getter and setter functions (but they can obviously be removed if the member variables are public). This is according to the [style guide](https://github.com/R2D2-2019/R2D2-2019/wiki/Style-Guide#structs-vs-classes).

- The constructor is, in my point of view, necessary to make it easier for the user to add certain gasses of the MQ-type of sensor. Through this construction, you only have to add the correct constructor line in the list of switch cases. In the main file, you will only have to add the enumeration to construct the entire object.

- The values in the array are **not** x and y of a certain scale or something. 
The **first** value is the first point on a logarithmic scale.
The **second** value is the last point on a logarithmic  scale.
The **third** value is the increment/decrement step on that specific logarithmic scale of that specific gas.

I think the previously mentioned items were a bit confusing to code reviewers. I hope this makes it a bit clearer.

Reply:
@olivierverwoerd Please edit your comment to reflect the explanation that @crisvdn gave.

Something like:
```C++
/** Each gas has its own logarithmic curve.
* The first value is the first point on a logarithmic scale
* The second value is the last point on a logarithmic scale
* the 3rd value is increment/decrement step on that specific scale.
**/
```

The current comment isn't accurate enough for someone to take over this class and use it.

Reply:
The first two comments of @crisvdn are fine. The last one is simply incorrect. It os just an starting point with a x and a y. NOT the last point.

I added the fact thats its logarithmic to doxygen. It will come by once its passed internal review.

Reply:
Can anyone link a source for this? At the moment it is @olivierverwoerd  's word against @crisvdn 's.
When this has been settled we can merge the PR

Reply:
I think [this](https://forum.arduino.cc/index.php?topic=469459.0) forum explains best.

> Should anyone stumble across this post in the future, this is how I "reversed engineered" the curve algorithm that is used in the sandbox electronics sketch. e.g. LPGCurve[3] =  {2.3,0.21,-0.47}. LOG(200) will give you 2.3. I used the AntiLog or Inverse Logarithm equation to get the value of 0.21, which plots near enough on the graph (1.6218). Its simple to work out the linear slope (-0.47) to complete the curve parameters.

Reply:
This comment is getting a bit out of hand here since I am a code reviewer and I originally designed what has come to be this class I want to give my opinion in the matter and clarify some things. When I started with gas detection I copied a bit of Arduino code since I was and still am very lazy, however since someone made, what was first an array, a class for the gas curve it would be much nicer to store the slope (third value of the array) in a variable which was suggested by @florianhumblot.

So to clarify what the array contains, it contains the start x, y and the slope. With all these values we can plot a line similar to the one on the datasheet. The way to calculate the slope is as followed [https://cdn.discordapp.com/attachments/553506933706063885/566259831804002304/IMG_20190412_155408493.jpg](url) And I have typed some explanation about calculating this in a document found here [https://drive.google.com/open?id=12-Kj6m5XBvGufl1lwmXAEi3uWt7KUJdR](url) (this was not previously added to the wiki or to the drive for some reason) I hope this clarifies the calculations. I advise to use option 2 suggested by @florianhumblot however since this is a design choice it

Reply:
Id say its good as it is now. Then we can focus on RobOS and get that done before Wednesday.